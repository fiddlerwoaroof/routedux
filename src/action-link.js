const ActionLink = (React, PropTypes) => {
  class ActionLink extends React.Component {
    constructor(props) {
      super(props);
    }
    render() {
      const { action, children, ...props } = this.props;

      const renderedRoute = this.store.pathForAction(action);

      return (
        <a
          href={renderedRoute}
          onClick={ev => {
            ev.preventDefault();
            this.store.dispatch(action);
          }}
          {...props}
        >
          {children}
        </a>
      );
    }
  }

  ActionLink.propTypes = {
    action: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
    children: PropTypes.node,
  };

  return ActionLink;
};

let OutComponent = ActionLink;
try {
  const React = require("react");
  const PropTypes = require("prop-types");

  OutComponent = ActionLink(React, PropTypes);
} catch (e) {
  /* empty */
}

export const _internal = {
  ActionLink,
};

export default OutComponent;
