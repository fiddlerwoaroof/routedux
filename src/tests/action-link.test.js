import { _internal } from "../action-link";
import Enzyme, { mount } from "enzyme";
import ezJson from "enzyme-to-json";
import Adapter from "@cfaester/enzyme-adapter-react-18";
Enzyme.configure({ adapter: new Adapter() });

import React from "react";
import PropTypes from "prop-types";

const _Link = _internal.ActionLink(React, PropTypes);

it("dispatches an action on click", () => {
  // given
  const store = {
    pathForAction: jest.fn(() => "/my/path"),
    dispatch: jest.fn(),
  };
  const props = {
    action: { type: "ACTION", id: "123" },
    children: "Hello World!",
  };
  class Link extends _Link {
    constructor() {
      super();
      this.store = store;
    }
  }

  const wrapper = mount(<Link {...{ ...props, store }} />);
  // when
  wrapper.simulate("click");

  //then
  expect(store.pathForAction.mock.calls).toEqual([
    [{ type: "ACTION", id: "123" }],
  ]);
  expect(store.dispatch.mock.calls).toEqual([[{ type: "ACTION", id: "123" }]]);
});

it("renders the url calculated by our internal function", () => {
  // given
  const store = {
    pathForAction: jest.fn(() => "/my/path"),
    dispatch: jest.fn(),
  };
  const props = {
    action: {},
    children: "Hello World!",
  };

  class Link extends _Link {
    constructor() {
      super();
      this.store = store;
    }
  }

  const wrapper = mount(<Link {...{ ...props, store }} />);

  expect(ezJson(wrapper)).toMatchSnapshot();
});

it("additional props are passed through", () => {
  // given
  const store = {
    pathForAction: jest.fn(() => "/my/path"),
    dispatch: jest.fn(),
  };
  const props = {
    action: {},
    children: "Hello World!",
    className: "foo",
  };

  class Link extends _Link {
    constructor() {
      super();
      this.store = store;
    }
  }

  const wrapper = mount(<Link {...{ ...props, store }} />);

  expect(ezJson(wrapper)).toMatchSnapshot();
});
