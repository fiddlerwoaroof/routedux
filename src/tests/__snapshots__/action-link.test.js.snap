// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`additional props are passed through 1`] = `
<Link
  action={{}}
  className="foo"
  store={
    {
      "dispatch": [MockFunction],
      "pathForAction": [MockFunction] {
        "calls": [
          [
            {},
          ],
        ],
        "results": [
          {
            "type": "return",
            "value": "/my/path",
          },
        ],
      },
    }
  }
>
  <a
    className="foo"
    href="/my/path"
    onClick={[Function]}
    store={
      {
        "dispatch": [MockFunction],
        "pathForAction": [MockFunction] {
          "calls": [
            [
              {},
            ],
          ],
          "results": [
            {
              "type": "return",
              "value": "/my/path",
            },
          ],
        },
      }
    }
  >
    Hello World!
  </a>
</Link>
`;

exports[`renders the url calculated by our internal function 1`] = `
<Link
  action={{}}
  store={
    {
      "dispatch": [MockFunction],
      "pathForAction": [MockFunction] {
        "calls": [
          [
            {},
          ],
        ],
        "results": [
          {
            "type": "return",
            "value": "/my/path",
          },
        ],
      },
    }
  }
>
  <a
    href="/my/path"
    onClick={[Function]}
    store={
      {
        "dispatch": [MockFunction],
        "pathForAction": [MockFunction] {
          "calls": [
            [
              {},
            ],
          ],
          "results": [
            {
              "type": "return",
              "value": "/my/path",
            },
          ],
        },
      }
    }
  >
    Hello World!
  </a>
</Link>
`;
